docs/design/unified_sidecar_and_process_monitor.md

```diff
@@ -0,0 +1,123 @@
+
+We will build a new process called `fdb-kubernetes-monitor` that is designed to run fdbserver processes itself. It will take a configuration file that specifies the processes to launch, including explicit environment variable substitutions, and will not have any mechanism for fdbmonitor conf file management. It will also have APIs to copy files into another folder, and to check the hashes of files, like the sidecar does. This process will not be a drop-in replacement for either the sidecar or fdbmonitor, and will be designed specifically for use within the Kubernetes operator. It will be available to use for other purposes, such as injecting FDB client libraries into a client application.
+
+fdb-kubernetes-monitor will have APIs for loading new configuration from the local copy of the config map, but will not automatically relaunch processes with new configuration. Instead, it will rely on having the operator do a synchronized kill so that we can instantly bounce all of the processes in a coordinated fashion. When fdb-kubernetes-monitor detects that a child process has been killed, it will launch a new child process with the latest configuration. It will also do appropriate signal handling for SIGKILL and SIGTERM signals in accordance with Docker best practices. The configuration file will be designed such that all of the processes of a single machine class can share a configuration file, with environment variables for command-line arguments that vary between processes. fdb-kubernetes-monitor will have a command-line option for how many fdbserver processes to spawn, and the configuration file will support defining arguments that vary between different processes in the pod based on the process number.
+
+During upgrades, we will have a sidecar container that we upgrade to the new version, with instructions to copy the binary into a shared volume. We will then update the configuration for fdb-kubernetes-monitor in the main container to tell it to use a newer version, which will cause it to use the version-scoped binary in the shared volume. We will include the sidecar container all the time in the initial work for this feature. In the future, we will explore an option to deploy it as an ephemeral container, once that feature is sufficiently stable in Kubernetes.
+
+fdb-kubernetes-monitor will log its activity to standard output, and will have an option to also log to a file.
+
+fdb-kubernetes-monitor will ship in an image called `foundationdb-kubernetes`. This image will be versioned with FoundationDB, and will include multiple client libraries, like the `foundationdb-kubernetes-sidecar` image currently does. The first `foundationdb-kubernetes` image we cut for a release will have the FoundationDB version as its tag. For subsequent updates to the image that do not involve a FoundationDB update, we will append a build number, of the format `-2`. For instance, we would have images in a series like `6.3.0`, `6.3.0-2`, `6.3.0-3`, and so on.
```

Comment:
How does the release process for `foundationdb-kubernetes` image work with FDB server release process pipeline?

How well these changes work with ongoing FDB server CI/CD work?

Reply:
I think it will be basically the same as the `foundationdb-kubernetes-sidecar` image that we build today. The pipeline that produces the `foundationdb` image will also have to produce the `foundationdb-kubernetes` image, and we will need an option to produce a new build of the latter image without rebuilding the former image.